modules/dnn/src/layers/attention_layer.cpp

```diff
@@ -0,0 +1,257 @@
+
+        CV_CheckEQ(input_shape[2], weight_shape[0], "DNN/Attention: invalid input shape");
+        CV_CheckEQ(static_cast<size_t>(weight_shape[1]), hidden_size, "DNN/Attention: invalid weight shape");
+        CV_CheckEQ(dim_bias, hidden_size, "DNN/Attention: invalid bias shape");
+
+        outputs.assign(1, inputs[0]);
+        return false;
+    }
+
+    virtual void finalize(InputArrayOfArrays inputs_arr, OutputArrayOfArrays outputs_arr) CV_OVERRIDE {
+        opt.init();
+
+        std::vector<Mat> inputs;
+        inputs_arr.getMatVector(inputs);
+        const auto input_shape = shape(inputs[0]);
+        batch_size = static_cast<size_t>(input_shape[0]);
+        seq_len = static_cast<size_t>(input_shape[1]);
+        input_hidden_size = static_cast<size_t>(input_shape[2]);
+    }
+
+    void forward(InputArrayOfArrays inputs_arr, OutputArrayOfArrays outputs_arr, OutputArrayOfArrays internals_arr) CV_OVERRIDE {
+        CV_TRACE_FUNCTION();
+        CV_TRACE_ARG_VALUE(name, "name", name.c_str());
+
+        if (inputs_arr.depth() == CV_16S)
+        {
+            forward_fallback(inputs_arr, outputs_arr, internals_arr);
+            return;
+        }
+
+        std::vector<Mat> inputs, outputs;
+        inputs_arr.getMatVector(inputs);
+        outputs_arr.getMatVector(outputs);
+
+        // prepack weights
+        if (!is_prepacked) {
+            // prepack
+            const auto &weight = inputs[1];
+            const auto *weight_data = weight.ptr<const float>();
+            packWeight(num_heads, qkv_head_sizes[0], input_hidden_size, weight_data,                                             hidden_size, packed_weight_q, opt);
+            packWeight(num_heads, qkv_head_sizes[1], input_hidden_size, weight_data + qkv_hidden_sizes[0],                       hidden_size, packed_weight_k, opt);
+            packWeight(num_heads, qkv_head_sizes[2], input_hidden_size, weight_data + qkv_hidden_sizes[0] + qkv_hidden_sizes[1], hidden_size, packed_weight_v, opt);
+
+            is_prepacked = true;
+        }
+
+        float *packed_weights[3] = {packed_weight_q.data(), packed_weight_k.data(), packed_weight_v.data()};
+        size_t packed_weights_size[3] = {packed_weight_q.size() / num_heads, packed_weight_k.size() / num_heads, packed_weight_v.size() / num_heads};
+
+        Mat gemm_buffer = Mat::zeros(1, int(batch_size * seq_len * hidden_size), CV_32F);
+        auto *Q = gemm_buffer.ptr<float>();
+        auto *K = Q + batch_size * seq_len * qkv_hidden_sizes[0];
+        auto *V = K + batch_size * seq_len * qkv_hidden_sizes[1];
+        float *QKV[3] = {Q, K, V}; // Q, K, V: [B, N, S, H]
+        {
+            const auto &input = inputs[0];
+            const auto &bias = inputs[2];
+            const auto *input_data = input.ptr<const float>();
+            const auto *bias_data = bias.ptr<const float>();
+
+            opt.multi_thread = false;
```

Comment:
Is there any problems with multiple threads?

Reply:
No, it is intended to do so.

We have input data of shape `[B, S, D]` and weight `[D, W]`, where `W = qk_hidden_size + qk_hidden_size + v_hidden_size` (also `qk_hidden_size = qk_head_size(H) * num_heads(N)`, `v_hidden_size = v_head_size * num_heads`, and `D=W` in most vits I observe), then weight is extraced repactively (with shape `[D, NH]`) and prepacked to have shape `[N, D, H]`. We loop over dimension `B` and `N` in parallel, so it turns into Gemm with two tensors of shape `[S, D]` and `[D, H]` for each iteration. This inner Gemm needs to be run in single-thread way.

Of course I know it is not the most efficient because usually `B=1` and `N` is also small (`N=12` in ViTs), while `S=197` and `D=768`. It can be further improved if time is enough before release.